core/services/relay/evm/functions/logpoller_wrapper.go

```diff
@@ -44,6 +79,153 @@ func (l *logPollerWrapper) Ready() error {
 	return nil
 }
 
-func (l *logPollerWrapper) LatestRoutes() (activeCoordinator common.Address, proposedCoordinator common.Address, err error) {
-	return common.Address{}, common.Address{}, nil
+// "public" methods of LogPollerWrapper
+func (l *logPollerWrapper) LatestRequests() ([]evmRelayTypes.OracleRequest, error) {
+	// TODO poll both active and proposed coordinators for requests and parse them
+
+	// logs, err := lp.Logs(
+	// 	end-lookbackBlocks,
+	// 	end,
+	// 	iregistry21.IKeeperRegistryMasterUpkeepPerformed{}.Topic(),
+	// 	registryAddress,
+	// 	pg.WithParentCtx(ctx),
+	// )
+	// if err != nil {
+	// 	return nil, fmt.Errorf("%w: failed to collect logs from log poller", err)
+	// }
+
+	// TODO:
+	// format logs
+
+	return nil, nil
+}
+
+func (l *logPollerWrapper) LatestResponses() ([]evmRelayTypes.OracleResponse, error) {
+	// TODO poll both active and proposed coordinators for responses and parse them
+
+	// logs, err := lp.Logs(
+	// 	end-lookbackBlocks,
+	// 	end,
+	// 	iregistry21.IKeeperRegistryMasterUpkeepPerformed{}.Topic(),
+	// 	registryAddress,
+	// 	pg.WithParentCtx(ctx),
+	// )
+	// if err != nil {
+	// 	return nil, fmt.Errorf("%w: failed to collect logs from log poller", err)
+	// }
+
+	// TODO:
+	// format logs
+
+	return nil, nil
+}
+
+// "internal" methods called within EVM relayer
+func (l *logPollerWrapper) SubscribeToUpdates(name string, subscriber evmRelayTypes.RouteUpdateSubscriber) {
+	if l.pluginConfig.ContractVersion == 0 {
+		// in V0, immediately set contract address to Oracle contract and never update again
+		_ = subscriber.UpdateRoutes(l.contractAddress, l.contractAddress)
+	}
+	// TODO subscribe
+}
+
+func (l *logPollerWrapper) checkForRouteUpdate() {
+	defer l.shutdownWaitGroup.Done()
+	freqSec := l.pluginConfig.ContractUpdateCheckFrequencySec
+	if freqSec == 0 {
+		l.lggr.Errorw("ContractUpdateCheckFrequencySec must set to more than 0 in PluginConfig")
+		return
+	}
+	ticker := time.NewTicker(time.Duration(freqSec) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-l.chStop:
+			return
+		case <-ticker.C:
+			// TODO: get Logs from Router
```

Comment:
what does that mean?

Reply:
lp.Logs() for FunctionsRouter.ContractProposed or  FunctionsRouter.ContractUpdated